rhizome/common/lib/util.rb

```diff
@@ -60,18 +60,36 @@ def sync_parent_dir(f)
 
 def safe_write_to_file(filename, content = nil)
   raise ArgumentError, "must provide either content or block" if (content.nil? && !block_given?) || (!content.nil? && block_given?)
-  temp_filename = filename + ".tmp"
-  lock_filename = "/tmp/#{Digest::SHA256.hexdigest(temp_filename)}.lock"
-  File.open(lock_filename, File::RDWR | File::CREAT) do |lock_file|
-    lock_file.flock(File::LOCK_EX)
+  atomic_file_replace(filename) do |_, new_file|
     if block_given?
-      File.open(temp_filename, File::RDWR | File::CREAT) do |f|
-        yield f
-      end
+      yield new_file
     else
-      File.write(temp_filename, content)
+      new_file.write(content)
+    end
+    nil
+  end
+end
+
+def atomic_file_replace(path)
+  lock_path = "/tmp/#{Digest::SHA256.hexdigest(path)}.lock"
```

Comment:
Not that we expect foul play here, but this isn't a secure approach to creating a temporary file against an attacker with access to the file system, because if you know the path, you can calculate the SHA256 and setup a symlink to a file they control.  You can to use `File::EXCL` in the flags below, so that the code raises if the file already exists.  That would probably remove the need for the `flock`.  You would then need to handle exceptions raised if the file exists, so that concurrent processes wait until this process finishes.  I think that seems like a better approach than using `flock`.

I don't have time to review the other parts of this for at least a week, as I'm about to board a flight to Japan.

Reply:
I don't think that's necessary for this patch to do though, since this patch has application goals in mind, and was following with the current implementation for the lock file. We can reconsider the lock file separately